### PR TITLE
Allow RF greater than number of zones to select more than one instance per zone

### PR DESCRIPTION
**Pull request by justinjung04**
**What this PR does**:

`ring.Get` that should return n (or more) instances which form the replicas for the given key uses the following configs:
- `ring.cfg.ReplicationFactor` to control how many replicas you would like to use per key
- `ring.cfg.ZoneAwarenessEnabled` to allow replicas to be across the zones evenly

Problem is, when `ZoneAwarenessEnabled` is enabled and `ReplicationFactor` is greater than number of available zones, the function was silently ignoring the ReplicationFactor and returned only one instance from each zone:

https://github.com/cortexproject/cortex/blob/a307fc04ce295391f8927da19299bb5d26ecfbd8/pkg/ring/ring.go#L376-L380

RF doesn't need to be bound to number of zones, and the following scenario is valid as well:
- There are 3 available zones
- Setting RF to 6 would choose 2 instances from each zone

This PR fixes the function to be able to return max `ceil(RF/len(zones))` number of instances per zone. For instance, numOfZones = 3 and RF = 4 will return 4 instances, and each zone will have either 1 or 2 instances returned.

**Which issue(s) this PR fixes**:
n/a

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ring/ring_test.go [hunk 1]
```diff
@@ -2020,7 +2030,7 @@ func TestRing_Get_NoMemoryAllocations(t *testing.T) {
 	buf, bufHosts, bufZones := MakeBuffersForGet()
 	r := rand.New(rand.NewSource(time.Now().UnixNano()))
 
-	numAllocs := testing.AllocsPerRun(10, func() {
+	numAllocs := testing.AllocsPerRun(100, func() {
```
Comment (harry671003) [inline pkg/ring/ring_test.go:2033]: Is this change required?

File: pkg/ring/ring_test.go [hunk 2]
```diff
@@ -447,26 +454,29 @@ func TestRing_Get_ZoneAwareness(t *testing.T) {
...
 					for i := 0; i < len(set.Instances); i++ {
-						if _, ok := zones[set.Instances[i].Zone]; ok {
-							t.Fatal("found multiple instances in the same zone")
+						if _, ok := zones[set.Instances[i].Zone]; !ok {
+							zones[set.Instances[i].Zone] = 1
+						} else {
+							zones[set.Instances[i].Zone]++
+
+							if zones[set.Instances[i].Zone] > int(maxNumOfHostsPerZone) {
+								t.Fatal("instances not spread across zones evenly")
```
Comment (yeya24) [inline pkg/ring/ring_test.go:477]: Can we add assertions to also make sure when RF > zones, each zone can get at least 1 instance?

File: pkg/ring/ring_test.go (hunk 1 above)
Comment (justinjung04) [reply pkg/ring/ring_test.go:2033]: Nope, will remove it :)

File: pkg/ring/ring_test.go (hunk 2 above)
Comment (justinjung04) [reply pkg/ring/ring_test.go:477]: Sure!

**Review by yeya24 — COMMENTED** [review]
Thanks, this looks good!

**Review by yeya24 — APPROVED** [review]
